test/WebSites/TempDataWebSite/Controllers/HomeController.cs

```diff
@@ -17,5 +18,30 @@ public IActionResult DisplayTempData(string value)
             TempData["key"] = value;
             return View();
         }
+
+        public IActionResult SetTempDataAndRedirect(string value)
+        {
+            TempData["key"] = value;
+            return RedirectToAction("GetTempData");
+        }
+
+        public string GetTempData()
+        {
+            var value = TempData["key"];
+            return value.ToString();
+        }
+
+        public IActionResult SetTempDataAndRedirectToPeek(string value)
+        {
+            TempData["key"] = value;
+            return RedirectToAction("PeekTempData");
+        }
+
+        public IActionResult PeekTempData()
+        {
+            var peekValue = TempData.Peek("key");
+            var getValue = TempData["key"];
```

Comment:
Indeed. The "key" test is this:

1. First request comes in, sets some TempData, does a redirect (or whatever)
2. Second request gets the TempData, confirms it is valid
3. Third request tries to get the TempData again, ensures that there *isn't* any TempData

That's the tricky thing with TempData: it takes three requests to verify the main scenario.